Nexus 6P / Arch Linux: failure to generate verity_key.pub

First, the issue, in brief:

When running ./run_all.sh, the install finishes with partial success, but throws a few errors to stderr.  Nexus 6P device can boot into Copperhead successfully, but not a single other feature of this package is installed (no root, gapps, orwall, etc. Just vanilla CopperheadOS.) There seems to be an issue with generating the verity_key.pub file, judging by the command outputs in bash.

Next, install details:

here is `uname -a`
`Linux 3p1cpwn4g3 4.8.13-1-ARCH #1 SMP PREEMPT Fri Dec 9 07:24:34 CET 2016 x86_64 GNU/Linux`

Here is `fastboot --version`
`fastboot version 7.0.0_r21`

This version supports flashing unlock, as per the instructions in the Readme.

Here is `gcc --version`
`gcc (GCC) 6.2.1 20160830 Copyright (C) 2016 Free Software Foundation, Inc. This is free software; see the source for copying conditions. There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.`

Here is `cpio --version`
`cpio (GNU cpio) 2.12 Copyright (C) 2015 Free Software Foundation, Inc. License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>. This is free software: you are free to change and redistribute it. There is NO WARRANTY, to the extent permitted by law. Written by Phil Nelson, David MacKenzie, John Oleynick, and Sergey Poznyakoff.`

Git was used to clone the repository, and Arch Linux uses git already for AUR, so for sure that is installed and working properly.

jre7-openjdk and jre8-openjdk are both installed. `archlinux-java status` gives
`Available Java environments: java-7-openjdk java-8-openjdk (default)`

also, here is `openssl version`
OpenSSL 1.0.2j 26 Sep 2016

A quick `sdiff` of the file list at 
https://www.archlinux.org/packages/core/x86_64/openssl/files/
against the file list at 
http://packages.ubuntu.com/precise/i386/libssl-dev/filelist
shows that the Arch Linux package 'openssl' contains all of the files in the Ubuntu 'libssl-dev' package.


The issue in full:
Following the instructions in the README, I am using a Nexus 6P.
Fastboot and adb are installed, versions are listed above.
Java JRE/JDK 1.7+ requirement is met (see above, JRE is java-8-openjdk by default)
git, cpri, gcc (g++) are all installed.  As a note, to avoid confusion, it would be better to simply list gcc, because g++ IS gcc.  (See https://gcc.gnu.org/onlinedocs/gcc-4.0.3/gcc/G_002b_002b-and-GCC.html for more information.)
udev is configured to allow running fastboot and adb as non-root.
OEM Unlocking is enabled on the Nexus 6P.
The factory image of CopperheadOS release was downloaded from https://copperhead.co/android/downloads, version NMF26Q.2016.12.25.22.27.39
Successfully got key for Daniel Micay and verified good signature on the factory image tarball.

Next, the script run_all.sh is run.
The same results as below can be observed whether using 'torsocks ./run_all.sh' or simply running the script on its own without torifying.  I have a log file of the bash command outputs for the entire install using both torsocks and bare script.  They are long, but if someone wants to comb through them, say the word.

The script continues installing all the way through flashing the CopperheadOS images, but throws the following errors as it progresses.  

`make: *** No rule to make target 'img2simg.c$', needed by 'img2simg'.  Stop.`
The script does not stop at this point, but continues the process, beginning with retrieving packages, and going on to verifying signatures with gpg.

gpg throws the following error at each signature verification (looks like three verifications, and this error is thrown three times)
`gpg: WARNING: unsafe permissions on homedir '/home/realname/Pets/mip/mission-improbable/packages'`
That said, the signatures still verify just fine, all good signatures by F-Droid.  Not sure if this is really part of the issue.

After verifying those three signatures, just before the pause in the script to allow the user to reboot into bootloader, the following error is thrown:
`./make_keys.sh: line 6: /home/realname/Pets/mip/mission-improbable/helper-repos/android-simg2img/generate_verity_key: No such file or directory`
The script then proceeds with 
`creating releasekey.pk8 with no password`
on the next line, before pausing for user input.

The line referenced by the error in the make_keys.sh script is 

`../extras/make_key verity '/C=CA/ST=Ontario/L=Toronto/O=CopperheadOS/OU=CopperheadOS/CN=CopperheadOS/emailAddress=copperheados@copperhead.co'
$SIMG2IMG_DIR/generate_verity_key -convert verity.x509.pem verity_key`

However the only file of the name 'generate_verity_key' in the directory indicated by SIMG2IMG_DIR/ is 'generate_verity_key.c'
I'm guessing there is actually supposed to be a second file called simply 'generate_verity_key.'  This is missing, and I haven't figured out why.

Unlocking bootloader is next, this proceeds with no errors.
Flashing images is next, this also proceeds with no errors.

At the end of flashing, I get
`Copperhead successfully installed!
~/Pets/mip/mission-improbable/angler-nmf26q ~/Pets/mip/mission-improbable
~/Pets/mip/mission-improbable
cp: cannot stat 'keys/verity_key.pub': No such file or directory`

At this point, the device is still in bootloader, unlocked.  I can start the device, and CopperheadOS boots successfully, just with absolutely no other parts of this package installed.



Hope this helps, let me know if you need other stuff to help fix the issue, I'd really like to try out this setup!

As an asside, is there a way to just run the scripts that run_all.sh calls individually?  The README seems to indicate this can be done, but no order of operations is clearly indicated (i.e. in what order are these scripts supposed to be run?)  Perhaps going through them all in order will show more clearly where the problem lies.  

Thanks for all the hard work on this, and Happy New Year!!

--- a/extras/build_verity_metadata.py
+++ b/extras/build_verity_metadata.py
@@ -1,5 +1,7 @@
 #! /usr/bin/env python
 
+from __future__ import print_function
+
 import os
 import sys
 import struct
@@ -13,7 +15,7 @@ METADATA_SIZE = BLOCK_SIZE * 8
 
 def run(cmd):
     status, output = commands.getstatusoutput(cmd)
-    print output
+    print(output)
     if status:
         exit(-1)
 
@@ -33,7 +35,7 @@ def sign_verity_table(table, signer_path, key_path):
             table_file.write(table)
             table_file.flush()
             cmd = " ".join((signer_path, table_file.name, key_path, signature_file.name))
-            print cmd
+            print(cmd)
             run(cmd)
             return signature_file.read()
 
@@ -63,7 +65,7 @@ def build_verity_metadata(data_blocks, metadata_image, root_hash,
 
 if __name__ == "__main__":
     if len(sys.argv) == 3 and sys.argv[1] == "-s":
-        print get_verity_metadata_size(int(sys.argv[2]))
+        print(get_verity_metadata_size(int(sys.argv[2])))
     elif len(sys.argv) == 8:
         data_image_blocks = int(sys.argv[1]) / 4096
         metadata_image = sys.argv[2]

--- a/extras/build_verity_metadata.py
+++ b/extras/build_verity_metadata.py
@@ -6,7 +6,8 @@ import os
 import sys
 import struct
 import tempfile
-import commands
+import subprocess
+import shlex
 
 VERSION = 0
 MAGIC_NUMBER = 0xb001b001
@@ -14,9 +15,14 @@ BLOCK_SIZE = 4096
 METADATA_SIZE = BLOCK_SIZE * 8
 
 def run(cmd):
-    status, output = commands.getstatusoutput(cmd)
-    print(output)
-    if status:
+    proc = subprocess.Popen(shlex.split(cmd),
+                            stdout=subprocess.PIPE,
+                            stderr=subprocess.PIPE)
+
+    print(proc.stderr.read())
+    print(proc.stdout.read())
+
+    if proc.returncode:
         exit(-1)
 
 def get_verity_metadata_size(data_size):

